### PR TITLE
Better redux actions proposal

### PR DESCRIPTION
**Pull request by agazso**
While working on the avatar caching I realized that often I tried to avoid using the app state because with the current setup it is very cumbersome to add a new action, especially when it's async:

In Actions.ts
N1.  You have to define an ActionType enum and an accompanying string (which is similar but not identical) as the identifier of the action
N2.  You have to create an action function that encapsulates the intent of state change and it can be dispatched later
N3.  You have to define an async action that will do the async functionality you want and ultimately will dispatch the state change

Then in index.ts
N4. You have to write in the appropriate reducer the change you want to apply on the state

Also, it's very cumbersome to find out what an action is doing:
In Component.tsx
F1.  For example you find the component where the action happens
In Container.ts
F2.  Then you have to go to the container of the component to see what is dispatched
In Action.ts
F3. Then you have to lookup the enum of the action, find the string name 
F4. If it's async it's more steps, because there is always an extra dispatch

In index.ts
F5. You have to find the reducer and see what happens

I propose a simplification so that it's no longer necessary to define reducer actions. This means that you can skip steps (N1 and N4) when creating new actions and (F3, F5) when finding out what an action does. Note that this means that you won't have to open `index.ts` when adding new actions or when you are figuring out what an action does.

I provided a helper action: `InternalActions.appStateUpdateWithFunction` receives an updater function that gets the state and returns a partial app state that will be `Object.apply`d. I had previous versions, where the you could pass a partial app state directly to a helper, which applied it on the state, but it has potential race condition issues with async actions, so I removed them.

I added a `timeTickWithoutReducer` reducer action which shows how to use, and an `updateAvatarPath` action that actually updates the previous state. It's not working correctly yet, because of some immutable-js issue that I could not figure out yet, but please review this PR to decide if it is the direction we want to follow in the future.

### CODE REVIEW

File: src/reducers/index.ts [hunk 1]
```diff
@@ -437,6 +435,13 @@ const appStateReducer = (state: AppState = defaultState, action: Actions): AppSt
             Debug.log('App state set');
             return action.payload.appState;
         }
+        case 'APP-STATE-UPDATE-FUNCTION': {
+            const stateCopy = {...state};
+            const partialAppState = action.payload.appStateUpdater(stateCopy);
+            Object.assign(stateCopy, partialAppState);
```
Comment (vujevits) [inline src/reducers/index.ts]: this could be {
   ...stateCopy,
   ...partialAppState,
}

**Comment by vujevits** [comment]
I like it!
Since this would _the_ way of updating the state (by convention), the naming should reflect that, now it sounds like there are other ways

File: src/reducers/index.ts (hunk 1 above)
Comment (agazso) [reply src/reducers/index.ts]: Or even 
```
return {
   ..state,
   ...partialAppState,
}
```

File: src/reducers/index.ts [hunk 2]
```diff
@@ -19,6 +19,7 @@ import { Debug } from '../Debug';
 import { PostFeed } from '../PostFeed';
 import { migrateAppState, currentAppStateVersion } from './migration';
 import { ModelHelper } from '../models/ModelHelper';
+import { string } from 'prop-types';
```
[the remark this answers is not in the document]
Comment (agazso) [reply src/reducers/index.ts]: Not at all! I think sometimes the VSCode plugin goes crazy and it starts doing things like this, one time it imported `console` from somewhere breaking logging completely...

[the remark this answers is not in the document]
Comment (vujevits) [reply src/reducers/index.ts]: shit happens, the first suggestion in the import list is often... useless.

**Comment by vujevits** [comment]
If it turns out good we can publish a blogpost about it ;-)

**Pull request merged**